src/framework/simple/mod.rs

```diff
@@ -0,0 +1,183 @@
+        self
+    }
+
+    pub fn delimiter(mut self, new_delimiter: &str) -> Self {
+        self.delimiter = new_delimiter.to_owned();
+        self
+    }
+
+    /// Stores all command names as lowercase strings, and converts to lowercase before checking for a command
+    pub fn case_insensitivity(mut self, case_insensitive: bool) -> Self {
+        self.case_insensitive = case_insensitive;
+        if self.case_insensitive {
+            let mut new_map = HashMap::new();
+            let command_iter = self.commands.drain();
+            for (k, v) in command_iter {
+                new_map.insert(k.to_lowercase(), v);
+            }
+            self.commands = new_map;
+        }
+        self
+    }
+
+    async fn run_cmd(&self, ctx: &Context, msg: &Message, cmd_name: &str, args: &[&str]) {
+        if let Some(cmd) = self.commands.get(cmd_name) {
+            if self.run_before_cmd(ctx, msg, cmd_name).await {
+                let res = cmd(ctx, msg, args).await;
+                self.run_after_cmd(ctx, msg, cmd_name, res).await;
+            }
+        } else if let Some(unrecongnized) = &self.default_cmd {
+            unrecongnized(ctx, msg, cmd_name, args).await;
+        }
+    }
+
+    /// returns true if the result of running the before fn is true, also returns true if there is no before fn set
+    async fn run_before_cmd(&self, ctx: &Context, msg: &Message, cmd_name: &str) -> bool {
+        if let Some(before) = &self.before_cmd {
+            before(ctx, msg, cmd_name).await
+        } else {
+            true
+        }
+    }
+
+    /// runs the after fn if it's set, no-op if there isn't one
+    async fn run_after_cmd(&self, ctx: &Context, msg: &Message, cmd_name: &str, res: CommandResult) {
+        if let Some(after) = &self.after_cmd {
+            after(ctx, msg, cmd_name, res).await;
+        }
+    }
+
+}
+
+#[async_trait]
+impl Framework for SimpleFramework {
+    async fn dispatch(&self, ctx: Context, msg: Message) {
+        if msg.content.starts_with(&self.prefix) {            
+            let cmd_and_args = msg.content.splitn(2, &self.prefix).collect::<Vec<&str>>();
+            if cmd_and_args[1].is_empty() {
+
+                if let Some(prefix_only) = &self.prefix_only_cmd {
+                    prefix_only(&ctx, &msg).await;
+                    return;
+                }
+
+            } else {
+                let tokens = cmd_and_args[1].split(&self.delimiter).collect::<Vec<&str>>();
+                
+                let cmd_name = if self.case_insensitive {
+                    tokens[0].to_lowercase()
+                } else {
+                    tokens[0].to_owned()
+                };
+                self.run_cmd(&ctx, &msg, &cmd_name, &tokens[1..]).await;
+
+            }
+
+        } else if let Some(normal_msg_fn) = &self.normal_message_fn {
+            normal_msg_fn(&ctx, &msg).await;
+        }
+        
+    }
+}
```

Comment:
The arguments could be handled better, but just to make a quick and efficient improvement to your implementation:
```suggestion
    if !msg.content.starts_with(&self.prefix) {        
        if let Some(normal_message) = &self.normal_message_fn {
            normal_message(&ctx, &msg).await;
        }
    
        return;
    }
    
    let cmd_and_args = &msg.content[self.prefix.len()..];
    
    if cmd_and_args.is_empty() {
        if let Some(prefix_only) = &self.prefix_only_cmd {
            prefix_only(&ctx, &msg).await;
        }
        
        return;
    }
    
    let tokens = cmd_and_args.split(&self.delimiter).collect::<Vec<&str>>();
    
    let cmd_name = if self.case_insensitive {
        Cow::Owned(tokens[0].to_lowercase())
    } else {
        Cow::Borrowed(&tokens[0])
    };
    
    self.run_cmd(&ctx, &msg, &cmd_name, &tokens[1..]).await;
```

This requires to import `Cow` as to avoid allocating the command name on every invocation if the user did not enable case insensitivity.

Reply:
I went ahead and rewrote to use the Args struct within the standard framework, and moved it to a shared module